In [1]:
# import libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.simplefilter('ignore')
warnings.filterwarnings('ignore')

In [34]:
from IPython.display import display
def display_full_dataframe(df):
    with pd.option_context('display.max_rows',None,'display.max_columns',None):
        display(df)

In [3]:
# upload the datasets
credit = pd.read_excel('Credit Banking_Project - 1.xls')

In [4]:
credit.columns

Index(['Sl No:', 'Customer', 'Age', 'City', 'Credit Card Product', 'Limit',
       'Company', 'Segment'],
      dtype='object')

In [5]:
credit.shape

(100, 8)

In [6]:
credit.head()

Sl No: Customer        Age       City Credit Card Product   Limit Company  \
0       1       A1   0.928521  BANGALORE                Gold  500000      C1   
1       2       A2  35.534551   CALCUTTA              Silver  100000      C2   
2       3       A3  11.559307     COCHIN            Platimum   10000      C3   
3       4       A4  45.820278     BOMBAY            Platimum   10001      C4   
4       5       A5  69.663948  BANGALORE            Platimum   10002      C5   

         Segment  
0  Self Employed  
1   Salaried_MNC  
2   Salaried_Pvt  
3           Govt  
4  Normal Salary

In [7]:
# Set column as index of the dataframe
credit.set_index('Sl No:', inplace=True)

In [8]:
credit.head()

Customer        Age       City Credit Card Product   Limit Company  \
Sl No:                                                                      
1            A1   0.928521  BANGALORE                Gold  500000      C1   
2            A2  35.534551   CALCUTTA              Silver  100000      C2   
3            A3  11.559307     COCHIN            Platimum   10000      C3   
4            A4  45.820278     BOMBAY            Platimum   10001      C4   
5            A5  69.663948  BANGALORE            Platimum   10002      C5   

              Segment  
Sl No:                 
1       Self Employed  
2        Salaried_MNC  
3        Salaried_Pvt  
4                Govt  
5       Normal Salary

In [9]:
credit.tail()

Customer        Age        City Credit Card Product   Limit Company  \
Sl No:                                                                       
96          A96  29.631637     CHENNAI              Silver  100000     C19   
97          A97  20.611833  TRIVANDRUM            Platimum   10000     C20   
98          A98  40.538985    CALCUTTA            Platimum   10001     C21   
99          A99  21.588666    CALCUTTA            Platimum   10002     C22   
100        A100  23.607638      COCHIN              Silver  100000      C5   

              Segment  
Sl No:                 
96       Salaried_Pvt  
97               Govt  
98      Normal Salary  
99      Self Employed  
100      Salaried_MNC

In [10]:
credit.isnull().sum()

Customer               0
Age                    0
City                   0
Credit Card Product    0
Limit                  0
Company                0
Segment                0
dtype: int64

There is no null value present in the datasets.

In [11]:
credit.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 100 entries, 1 to 100
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Customer             100 non-null    object 
 1   Age                  100 non-null    float64
 2   City                 100 non-null    object 
 3   Credit Card Product  100 non-null    object 
 4   Limit                100 non-null    int64  
 5   Company              100 non-null    object 
 6   Segment              100 non-null    object 
dtypes: float64(1), int64(1), object(5)
memory usage: 6.2+ KB


In [12]:
credit.describe().T

count           mean            std           min            25%  \
Age    100.0      40.146505      23.629594      0.726493      20.254138   
Limit  100.0  284700.610000  218494.967852  10000.000000  100000.000000   

                 50%           75%            max  
Age        40.545682      61.60691      78.431756  
Limit  300001.500000  500000.00000  500000.000000

 The age value which is less than 18 is invalid data so I will repalce that value with mean i.e **40.14**

In [13]:
credit['Age']

Sl No:
1       0.928521
2      35.534551
3      11.559307
4      45.820278
5      69.663948
         ...    
96     29.631637
97     20.611833
98     40.538985
99     21.588666
100    23.607638
Name: Age, Length: 100, dtype: float64

In [14]:
mean_age = credit['Age'].mean()
print(mean_age)

40.14650487273873


In [15]:
credit.loc[(credit['Age']<18),'Age'] = mean_age

In [16]:
credit['Age']

Sl No:
1      40.146505
2      35.534551
3      40.146505
4      45.820278
5      69.663948
         ...    
96     29.631637
97     20.611833
98     40.538985
99     21.588666
100    23.607638
Name: Age, Length: 100, dtype: float64

In [17]:
new_mean_age = credit['Age'].mean()
print(new_mean_age)

47.28032183784049


The new mean of age column is **47.28** so we can continue with this approach as the datasets did not change its nature much.

In [18]:
# importing spend worksheet 
spend = pd.read_excel('Credit Banking_Project - 1.xls', sheet_name='Spend')

In [19]:
spend

Sl No: Customer      Month        Type         Amount
0          1       A1 2004-01-12   JEWELLERY  473775.834869
1          2       A1 2004-01-03       PETRO  335578.666019
2          3       A1 2004-01-15     CLOTHES  371040.941912
3          4       A1 2004-01-25        FOOD  141177.813256
4          5       A1 2005-01-17      CAMERA  398403.771812
...      ...      ...        ...         ...            ...
1495    1496      A67 2006-02-04  BUS TICKET  195841.312215
1496    1497      A68 2006-03-25  BUS TICKET  284262.517709
1497    1498      A69 2006-03-31  BUS TICKET  206552.273598
1498    1499      A70 2006-03-23  BUS TICKET   41024.781283
1499    1500      A71 2006-03-24  BUS TICKET  101397.548588

[1500 rows x 5 columns]

In [20]:
spend.set_index('Sl No:', inplace=True)

In [21]:
spend.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1500 entries, 1 to 1500
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Customer  1500 non-null   object        
 1   Month     1500 non-null   datetime64[ns]
 2   Type      1500 non-null   object        
 3   Amount    1500 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(2)
memory usage: 58.6+ KB


In [22]:
spend['S Month'] = spend['Month'].dt.strftime('%B')
spend['S Year'] = spend['Month'].dt.year

In [23]:
spend = spend.drop(columns=['Month'])

In [24]:
spend

Customer        Type         Amount   S Month  S Year
Sl No:                                                      
1            A1   JEWELLERY  473775.834869   January    2004
2            A1       PETRO  335578.666019   January    2004
3            A1     CLOTHES  371040.941912   January    2004
4            A1        FOOD  141177.813256   January    2004
5            A1      CAMERA  398403.771812   January    2005
...         ...         ...            ...       ...     ...
1496        A67  BUS TICKET  195841.312215  February    2006
1497        A68  BUS TICKET  284262.517709     March    2006
1498        A69  BUS TICKET  206552.273598     March    2006
1499        A70  BUS TICKET   41024.781283     March    2006
1500        A71  BUS TICKET  101397.548588     March    2006

[1500 rows x 5 columns]

In [25]:
# importing repayment worksheet
repayment = pd.read_excel('Credit Banking_Project - 1.xls', sheet_name='Repayment')

In [26]:
repayment

SL No: Customer      Month         Amount
0          1       A1 2004-01-12  331844.007374
1          2       A1 2004-01-03  441139.807335
2          3       A1 2004-01-15   32480.004009
3          4       A1 2004-01-25   90636.966534
4          5       A1 2005-01-17    1581.969829
...      ...      ...        ...            ...
1495    1496      A67 2005-05-09  200288.874065
1496    1497      A68 2006-05-10  231194.702295
1497    1498      A69 2006-07-11  391195.615836
1498    1499      A70 2005-08-12  357629.618186
1499    1500      A71 2004-09-13  179771.745248

[1500 rows x 4 columns]

In [27]:
repayment.set_index('SL No:', inplace=True)

In [28]:
repayment.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1500 entries, 1 to 1500
Data columns (total 3 columns):
 #   Column    Non-Null Count  Dtype         
---  ------    --------------  -----         
 0   Customer  1500 non-null   object        
 1   Month     1500 non-null   datetime64[ns]
 2   Amount    1500 non-null   float64       
dtypes: datetime64[ns](1), float64(1), object(1)
memory usage: 46.9+ KB


In [29]:
repayment['R Month'] = repayment['Month'].dt.strftime('%B')
repayment['R Year'] = repayment['Month'].dt.year
repayment = repayment.drop(columns=['Month'])

In [31]:
repayment = repayment.rename(columns={'Amount':'Repayment'})

In [32]:
repayment

Customer      Repayment    R Month  R Year
SL No:                                           
1            A1  331844.007374    January    2004
2            A1  441139.807335    January    2004
3            A1   32480.004009    January    2004
4            A1   90636.966534    January    2004
5            A1    1581.969829    January    2005
...         ...            ...        ...     ...
1496        A67  200288.874065        May    2005
1497        A68  231194.702295        May    2006
1498        A69  391195.615836       July    2006
1499        A70  357629.618186     August    2005
1500        A71  179771.745248  September    2004

[1500 rows x 4 columns]

## Customer who exceed the limit in any particluar month

In [33]:
merge_credit_spend = pd.merge(credit,spend,on='Customer',how='left')

In [35]:
display_full_dataframe(merge_credit_spend)

Customer        Age        City Credit Card Product   Limit Company  \
0          A1  40.146505   BANGALORE                Gold  500000      C1   
1          A1  40.146505   BANGALORE                Gold  500000      C1   
2          A1  40.146505   BANGALORE                Gold  500000      C1   
3          A1  40.146505   BANGALORE                Gold  500000      C1   
4          A1  40.146505   BANGALORE                Gold  500000      C1   
5          A1  40.146505   BANGALORE                Gold  500000      C1   
6          A1  40.146505   BANGALORE                Gold  500000      C1   
7          A1  40.146505   BANGALORE                Gold  500000      C1   
8          A1  40.146505   BANGALORE                Gold  500000      C1   
9          A1  40.146505   BANGALORE                Gold  500000      C1   
10         A1  40.146505   BANGALORE                Gold  500000      C1   
11         A1  40.146505   BANGALORE                Gold  500000      C1   
12         A1  40.146505   BANGALORE                Gold  500000      C1   
13         A1  40.146505   BANGALORE                Gold  500000      C1   
14         A1  40.146505   BANGALORE                Gold  500000      C1   
15         A1  40.146505   BANGALORE                Gold  500000      C1   
16         A1  40.146505   BANGALORE                Gold  500000      C1   
17         A1  40.146505   BANGALORE                Gold  500000      C1   
18         A2  35.534551    CALCUTTA              Silver  100000      C2   
19         A2  35.534551    CALCUTTA              Silver  100000      C2   
20         A2  35.534551    CALCUTTA              Silver  100000      C2   
21         A2  35.534551    CALCUTTA              Silver  100000      C2   
22         A2  35.534551    CALCUTTA              Silver  100000      C2   
23         A2  35.534551    CALCUTTA              Silver  100000      C2   
24         A2  35.534551    CALCUTTA              Silver  100000      C2   
25         A2  35.534551    CALCUTTA              Silver  100000      C2   
26         A2  35.534551    CALCUTTA              Silver  100000      C2   
27         A2  35.534551    CALCUTTA              Silver  100000      C2   
28         A2  35.534551    CALCUTTA              Silver  100000      C2   
29         A3  40.146505      COCHIN            Platimum   10000      C3   
30         A3  40.146505      COCHIN            Platimum   10000      C3   
31         A3  40.146505      COCHIN            Platimum   10000      C3   
32         A3  40.146505      COCHIN            Platimum   10000      C3   
33         A3  40.146505      COCHIN            Platimum   10000      C3   
34         A3  40.146505      COCHIN            Platimum   10000      C3   
35         A3  40.146505      COCHIN            Platimum   10000      C3   
36         A3  40.146505      COCHIN            Platimum   10000      C3   
37         A3  40.146505      COCHIN            Platimum   10000      C3   
38         A3  40.146505      COCHIN            Platimum   10000      C3   
39         A3  40.146505      COCHIN            Platimum   10000      C3   
40         A4  45.820278      BOMBAY            Platimum   10001      C4   
41         A4  45.820278      BOMBAY            Platimum   10001      C4   
42         A4  45.820278      BOMBAY            Platimum   10001      C4   
43         A4  45.820278      BOMBAY            Platimum   10001      C4   
44         A4  45.820278      BOMBAY            Platimum   10001      C4   
45         A4  45.820278      BOMBAY            Platimum   10001      C4   
46         A4  45.820278      BOMBAY            Platimum   10001      C4   
47         A4  45.820278      BOMBAY            Platimum   10001      C4   
48         A4  45.820278      BOMBAY            Platimum   10001      C4   
49         A4  45.820278      BOMBAY            Platimum   10001      C4   
50         A4  45.820278      BOMBAY            Platimum   10001      C4   
51         A5  69.663948   BANGALORE            Plati

In [36]:
merge_credit_spend['Exceed_limit'] = merge_credit_spend['Amount'] - merge_credit_spend['Limit']

In [37]:
limit = merge_credit_spend.groupby(['Customer','S Month','S Year'])['Exceed_limit'].sum()

In [38]:
limit = pd.DataFrame(limit)
limit

Exceed_limit
Customer S Month  S Year              
A1       April    2006   -4.354931e+05
         August   2005   -3.706117e+05
         February 2004   -4.586189e+05
                  2005   -5.958070e+05
         January  2004   -1.488827e+06
...                                ...
A95      January  2004   -1.521661e+05
A96      January  2004    2.203635e+05
A97      January  2004    1.543300e+05
A98      January  2004    7.748251e+04
A99      January  2004    4.660184e+05

[807 rows x 1 columns]

In [39]:
cross_limit = limit>0

In [41]:
display_full_dataframe(cross_limit)

Exceed_limit
Customer S Month   S Year              
A1       April     2006           False
         August    2005           False
         February  2004           False
                   2005           False
         January   2004           False
                   2005           False
         May       2004           False
         November  2005           False
         October   2006           False
A10      April     2006           False
         August    2006           False
         February  2005           False
         January   2004           False
         July      2006           False
         March     2004           False
                   2006           False
         May       2004           False
                   2005           False
                   2006           False
         November  2005           False
A100     January   2004           False
A11      April     2004           False
         August    2005           False
         February  2004           False
                   2005           False
         January   2004           False
         June      2005           False
                   2006           False
         March     2006           False
         May       2005           False
                   2006           False
         November  2006           False
         September 2006           False
A12      April     2005           False
                   2006           False
         August    2006           False
         February  2004           False
                   2005           False
         January   2004           False
                   2005           False
         July      2006           False
         June      2005           False
         March     2006           False
         May       2005           False
         November  2005           False
         October   2006           False
         September 2004           False
A13      April     2004           False
                   2006           False
         August    2006           False
         February  2005           False
                   2006           False
         January   2004           False
         July      2005           False
         June      2005           False
         March     2006           False
         May       2004           False
                   2005           False
                   2006           False
         November  2004           False
                   2005           False
                   2006           False
         September 2006           False
A14      April     2004           False
                   2005           False
                   2006           False
         August    2005           False
         December  2006           False
         February  2005           False
         January   2004           False
         July      2005           False
                   2006           False
         June      2005           False
         March     2006           False
         May       2005           False
                   2006           False
         November  2006           False
         October   2006           False
         September 2006           False
A15      August    2005           False
         February  2004           False
                   2005           False
                   2006           False
         January   2004           False
         July      2005           False
         June      2005           False
                   2006           False
         March     2006           False
         May       2005           False
                   2006           False
         November  2006           False
         September 2005           False
A16      August    2005           False
         December  2006           False
         February  2005           False
                   2006           False
         January   2004           False
         July      2006           False
         June      2005    

##  Monthly spend of each customer.

In [43]:
monthly_spend = spend.groupby(['Customer', 'S Month','S Year'])['Amount'].sum()

In [44]:
monthly_spend = pd.DataFrame(monthly_spend)

In [46]:
display_full_dataframe(monthly_spend)

Amount
Customer S Month   S Year              
A1       April     2006    5.645069e+05
         August    2005    1.293883e+05
         February  2004    4.138111e+04
                   2005    1.404193e+06
         January   2004    1.511173e+06
                   2005    3.984038e+05
         May       2004    1.311966e+05
         November  2005    4.573169e+05
         October   2006    2.207350e+05
A10      April     2006    6.021290e+04
         August    2006    2.308944e+05
         February  2005    3.573965e+05
         January   2004    7.474281e+05
         July      2006    2.674369e+05
         March     2004    4.351586e+05
                   2006    2.337414e+05
         May       2004    4.807288e+05
                   2005    6.960673e+05
                   2006    1.172956e+05
         November  2005    4.844257e+05
A100     January   2004    4.225419e+04
A11      April     2004    4.931042e+05
         August    2005    4.539811e+05
         February  2004    5.019403e+05
                   2005    4.926942e+05
         January   2004    8.195447e+05
         June      2005    4.127473e+05
                   2006    4.627846e+05
         March     2006    1.488703e+05
         May       2005    3.377852e+05
                   2006    1.039398e+05
         November  2006    2.616615e+05
         September 2006    9.287055e+04
A12      April     2005    5.244122e+05
                   2006    1.795846e+05
         August    2006    3.471292e+05
         February  2004    4.143921e+05
                   2005    1.208185e+06
         January   2004    7.449706e+05
                   2005    2.878571e+05
         July      2006    1.337266e+05
         June      2005    9.294122e+05
         March     2006    8.775819e+05
         May       2005    8.596220e+05
         November  2005    6.478689e+04
         October   2006    8.687326e+04
         September 2004    1.900529e+05
A13      April     2004    2.678509e+05
                   2006    5.929006e+05
         August    2006    1.446846e+05
         February  2005    2.032310e+06
                   2006    1.169444e+05
         January   2004    6.336622e+05
         July      2005    4.618902e+05
         June      2005    3.967439e+05
         March     2006    1.006233e+06
         May       2004    3.355591e+05
                   2005    5.986687e+05
                   2006    2.531827e+05
         November  2004    4.894623e+05
                   2005    3.967534e+05
                   2006    6.463962e+04
         September 2006    2.508526e+05
A14      April     2004    1.644898e+05
                   2005    4.696613e+05
                   2006    1.850607e+05
         August    2005    7.470524e+04
         December  2006    4.393573e+05
         February  2005    1.804303e+06
         January   2004    5.154660e+05
         July      2005    1.005640e+05
                   2006    4.805658e+05
         June      2005    8.190180e+04
         March     2006    1.161758e+06
         May       2005    5.259856e+05
                   2006    5.176778e+05
         November  2006    4.785365e+05
         October   2006    1.135759e+05
         September 2006    1.341510e+05
A15      August    2005    3.389989e+05
         February  2004    1.948092e+05
                   2005    2.025370e+05
                   2006    2.153967e+05
         January   2004    7.940466e+05
         July      2005    1.966673e+05
         June      2005    4.734607e+05
                   2006    3.454276e+05
         March     2006    7.746789e+05
         May       2005    6.254104e+05
                   2006    2.669103e+05
         November  2006    2.071003e+05
         September 2005    1.313575e+05
A16      August    2005    2.700026e+05
         December  2006    2.202590e+05
         February  2005    4.208625e+05
                   2006    1.649431e+05
         January   2004    7.220161e+05
         July      2006    8.985865e+05
         June      2005    1.6056

In [60]:
monthly_spend.max()

Amount    2.605015e+06
dtype: float64

In [61]:
monthly_spend.idxmax()

Amount    (A37, January, 2004)
dtype: object

##  Monthly repayment of each customer.

In [48]:
monthly_repayment = repayment.groupby(['Customer','R Month','R Year'])['Repayment'].sum()

In [49]:
monthly_repayment = pd.DataFrame(monthly_repayment)

In [50]:
monthly_repayment

Repayment
Customer R Month  R Year              
A1       April    2006    3.712733e+05
         February 2004    1.911800e+05
                  2005    1.199808e+06
         January  2004    1.362775e+06
                  2005    1.581970e+03
...                                ...
A95      January  2004    7.510949e+04
A96      January  2004    1.101390e+05
A97      January  2004    1.746064e+05
A98      January  2004    9.780260e+04
A99      January  2004    3.585899e+05

[798 rows x 1 columns]

In [64]:
monthly_repayment.max()

Repayment    3.026776e+06
dtype: float64

In [65]:
monthly_repayment.idxmax()

Repayment    (A41, January, 2004)
dtype: object

##  Highest paying 10 customers.

In [62]:
highest_10 = spend.groupby(['Customer','S Month', 'S Year'])['Amount'].sum().nlargest(10)

In [63]:
highest_10

Customer  S Month   S Year
A37       January   2004      2.605015e+06
A39       January   2004      2.417496e+06
A60       January   2004      2.245218e+06
A40       January   2004      2.233137e+06
A22       January   2004      2.126632e+06
A13       February  2005      2.032310e+06
A38       January   2004      1.991223e+06
A21       March     2006      1.963231e+06
A42       January   2004      1.940790e+06
A36       January   2004      1.905283e+06
Name: Amount, dtype: float64

##  People in which segment are spending more money

In [54]:
segment_spend = merge_credit_spend.groupby(['Segment','S Month','S Year'])['Amount'].sum()

In [55]:
segment_spend = pd.DataFrame(segment_spend)

In [57]:
display_full_dataframe(segment_spend)

Amount
Segment       S Month   S Year              
Govt          April     2004    4.709543e+05
                        2005    5.301301e+05
                        2006    4.537539e+06
              August    2005    1.514305e+06
                        2006    4.478368e+05
              December  2005    1.169054e+06
                        2006    2.710380e+05
              February  2004    7.841881e+05
                        2005    6.170115e+06
                        2006    3.956979e+06
              January   2004    1.195284e+07
                        2005    2.323803e+06
                        2006    2.958979e+05
              July      2005    4.644468e+05
                        2006    2.045870e+06
              June      2005    2.876111e+06
                        2006    3.454276e+05
              March     2004    5.294807e+05
                        2006    9.639553e+06
              May       2004    1.593277e+06
                        2005    5.899580e+06
                        2006    2.872176e+06
              November  2004    9.708668e+05
                        2005    1.418180e+06
                        2006    1.058487e+06
              October   2005    2.135328e+05
                        2006    8.037678e+05
              September 2004    4.123667e+05
                        2005    7.597169e+05
                        2006    9.981105e+05
Normal Salary April     2004    2.007181e+06
                        2005    1.705185e+06
                        2006    6.119545e+06
              August    2005    2.417785e+06
                        2006    9.807534e+05
              December  2005    5.664141e+05
                        2006    1.510012e+06
              February  2004    2.841896e+06
                        2005    9.404518e+06
                        2006    3.530645e+06
              January   2004    2.239460e+07
                        2005    1.743615e+06
                        2006    1.252828e+06
              July      2005    2.361115e+05
                        2006    2.130229e+06
              June      2005    2.244533e+06
                        2006    1.643947e+06
              March     2004    8.324581e+05
                        2006    1.687793e+07
              May       2004    8.486817e+05
                        2005    9.952877e+06
                        2006    5.510276e+06
              November  2004    1.431529e+06
                        2005    2.261307e+06
                        2006    2.485723e+06
              October   2005    1.487898e+06
                        2006    1.093331e+06
              September 2004    1.212598e+06
                        2005    7.731255e+05
                        2006    2.096079e+05
Salaried_MNC  April     2004    2.262437e+06
                        2005    4.980228e+05
                        2006    3.344838e+06
              August    2005    6.124006e+05
                        2006    1.446846e+05
              December  2005    1.001626e+06
                        2006    6.528509e+05
              February  2004    1.277256e+06
                        2005    6.089046e+06
                        2006    1.245771e+06
              January   2004    1.343229e+07
                        2005    1.067623e+06
              July      2005    1.557178e+06
                        2006    1.218794e+06
              June      2005    1.284365e+06
              March     2004    4.531288e+05
                        2006    9.220944e+06
              May       2004    7.315066e+05
                        2005    5.994409e+06
                        2006    4.269927e+06
              November  2004    1.318011e+06
                        2005    9.754088e+05
                        2006    1.451006e+06
              October   2005    7.787577e+05
                        2006    5.015692e+05
              September 2004    8.232033e+05
                        2005    4.448555e+05
                        2006    9

In [58]:
segment_spend.max()

Amount    2.239460e+07
dtype: float64

In [59]:
print(segment_spend.idxmax())

Amount    (Normal Salary, January, 2004)
dtype: object


## Which age group is spending more money?

In [68]:
age_range = [18,25,30,35,40,45,50,55,60,65,70,75,80]
merge_credit_spend['Age Group'] = pd.cut(merge_credit_spend['Age'], bins=age_range, labels=['18-25','25-30','30-35','35-40','40-45','45-50','50-55','55-60','60-65','65-70','70-75','75-80'])
age_group_spending =  merge_credit_spend.groupby(['Age Group','S Month', 'S Year'])['Amount'].sum()

In [72]:
age_group_spending = pd.DataFrame(age_group_spending)
display_full_dataframe(age_group_spending)

Amount
Age Group S Month   S Year              
18-25     April     2004    9.447752e+04
                    2005    3.914229e+05
                    2006    6.692370e+05
          August    2004    0.000000e+00
                    2005    4.086292e+05
                    2006    0.000000e+00
          December  2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          February  2004    5.238944e+05
                    2005    1.654514e+06
                    2006    8.276079e+05
          January   2004    5.264562e+06
                    2005    7.875854e+05
                    2006    0.000000e+00
          July      2004    0.000000e+00
                    2005    1.966673e+05
                    2006    8.437599e+05
          June      2004    0.000000e+00
                    2005    4.734607e+05
                    2006    9.952423e+05
          March     2004    0.000000e+00
                    2005    0.000000e+00
                    2006    2.329401e+06
          May       2004    0.000000e+00
                    2005    2.236076e+06
                    2006    8.906201e+05
          November  2004    0.000000e+00
                    2005    2.556433e+04
                    2006    2.071003e+05
          October   2004    0.000000e+00
                    2005    2.964740e+05
                    2006    0.000000e+00
          September 2004    3.248447e+05
                    2005    1.313575e+05
                    2006    4.725656e+05
25-30     April     2004    3.076356e+05
                    2005    8.183135e+05
                    2006    1.951631e+06
          August    2004    0.000000e+00
                    2005    3.799375e+05
                    2006    1.662091e+05
          December  2004    0.000000e+00
                    2005    3.683637e+05
                    2006    4.604023e+05
          February  2004    2.892789e+05
                    2005    8.355308e+05
                    2006    1.386880e+06
          January   2004    4.895163e+06
                    2005    8.092472e+05
                    2006    0.000000e+00
          July      2004    0.000000e+00
                    2005    4.623161e+05
                    2006    4.880537e+05
          June      2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          March     2004    0.000000e+00
                    2005    0.000000e+00
                    2006    4.743402e+06
          May       2004    3.515052e+05
                    2005    2.080551e+06
                    2006    1.470861e+06
          November  2004    4.161792e+05
                    2005    3.695698e+05
                    2006    8.789392e+05
          October   2004    0.000000e+00
                    2005    3.013661e+05
                    2006    2.720611e+05
          September 2004    4.852799e+05
                    2005    4.920385e+05
                    2006    5.629584e+05
30-35     April     2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          August    2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          December  2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          February  2004    0.000000e+00
                    2005    5.939075e+05
                    2006    0.000000e+00
          January   2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          July      2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          June      2004    0.000000e+00
                    2005    0.000000e+00
                    2006    0.000000e+00
          March     2004    0.000000e+00
                    2005    0.000000e+00
                    2006    4.102478e+04
          May   

In [73]:
age_group_spending.max()

Amount    2.246502e+07
dtype: float64

In [74]:
age_group_spending.idxmax()

Amount    (40-45, January, 2004)
dtype: object

In [76]:
merge = pd.merge(repayment,merge_credit_spend,on='Customer',how='left')

In [78]:
merge

Customer      Repayment    R Month  R Year        Age       City  \
0           A1  331844.007374    January    2004  40.146505  BANGALORE   
1           A1  331844.007374    January    2004  40.146505  BANGALORE   
2           A1  331844.007374    January    2004  40.146505  BANGALORE   
3           A1  331844.007374    January    2004  40.146505  BANGALORE   
4           A1  331844.007374    January    2004  40.146505  BANGALORE   
...        ...            ...        ...     ...        ...        ...   
37279      A69  391195.615836       July    2006  77.791501   CALCUTTA   
37280      A70  357629.618186     August    2005  34.869273     BOMBAY   
37281      A70  357629.618186     August    2005  34.869273     BOMBAY   
37282      A71  179771.745248  September    2004  52.333796     BOMBAY   
37283      A71  179771.745248  September    2004  52.333796     BOMBAY   

      Credit Card Product   Limit Company        Segment        Type  \
0                    Gold  500000      C1  Self Employed   JEWELLERY   
1                    Gold  500000      C1  Self Employed       PETRO   
2                    Gold  500000      C1  Self Employed     CLOTHES   
3                    Gold  500000      C1  Self Employed        FOOD   
4                    Gold  500000      C1  Self Employed      CAMERA   
...                   ...     ...     ...            ...         ...   
37279            Platimum   10000     C31  Self Employed  BUS TICKET   
37280            Platimum   10001     C32  Self Employed  AIR TICKET   
37281            Platimum   10001     C32  Self Employed  BUS TICKET   
37282            Platimum   10002     C33  Self Employed  AIR TICKET   
37283            Platimum   10002     C33  Self Employed  BUS TICKET   

              Amount   S Month  S Year   Exceed_limit Age Group  
0      473775.834869   January    2004  -26224.165131     40-45  
1      335578.666019   January    2004 -164421.333981     40-45  
2      371040.941912   January    2004 -128959.058088     40-45  
3      141177.813256   January    2004 -358822.186744     40-45  
4      398403.771812   January    2005 -101596.228188     40-45  
...              ...       ...     ...            ...       ...  
37279  206552.273598     March    2006  196552.273598     75-80  
37280  161440.544350  February    2005  151439.544350     30-35  
37281   41024.781283     March    2006   31023.781283     30-35  
37282  423694.651910  February    2005  413692.651910     50-55  
37283  101397.548588     March    2006   91395.548588     50-55  

[37284 rows x 16 columns]

In [79]:
merge.isnull().sum()

Customer               0
Repayment              0
R Month                0
R Year                 0
Age                    0
City                   0
Credit Card Product    0
Limit                  0
Company                0
Segment                0
Type                   0
Amount                 0
S Month                0
S Year                 0
Exceed_limit           0
Age Group              0
dtype: int64

## Which is the most profitable segment?

In [80]:
merge['Profit'] = merge['Amount'] - merge['Repayment']

In [84]:
Profitable_segment = merge.groupby(['Segment','S Month','S Year'])['Profit'].sum()

In [85]:
Profitable_segment = pd.DataFrame(Profitable_segment)
display_full_dataframe(Profitable_segment)

Profit
Segment       S Month   S Year              
Govt          April     2004   -9.788844e+06
                        2005   -5.237065e+06
                        2006   -2.338157e+06
              August    2005    2.050456e+06
                        2006   -5.979772e+05
              December  2005    1.802327e+06
                        2006   -9.049861e+06
              February  2004    1.248941e+06
                        2005   -7.092837e+05
                        2006    2.067387e+07
              January   2004   -3.150832e+07
                        2005    1.299144e+07
                        2006    1.478566e+06
              July      2005   -1.361748e+07
                        2006    1.518121e+07
              June      2005    2.455386e+07
                        2006    2.833129e+06
              March     2004    2.783715e+06
                        2006   -6.626402e+06
              May       2004    1.245165e+07
                        2005    2.938431e+07
                        2006    1.274579e+07
              November  2004   -8.036285e+05
                        2005    1.461478e+06
                        2006   -1.103764e+07
              October   2005   -2.420040e+06
                        2006    1.292740e+07
              September 2004   -4.079013e+06
                        2005   -1.579635e+05
                        2006    9.893412e+06
Normal Salary April     2004   -6.112233e+06
                        2005    2.262880e+06
                        2006    1.415686e+07
              August    2005    9.815917e+06
                        2006   -4.260136e+06
              December  2005   -8.157171e+06
                        2006    4.282882e+06
              February  2004    4.723485e+05
                        2005   -2.116723e+07
                        2006    8.421209e+06
              January   2004    4.511346e+07
                        2005   -7.407707e+05
                        2006   -5.178008e+06
              July      2005   -7.564717e+06
                        2006   -1.794642e+06
              June      2005   -7.907004e+06
                        2006    1.182325e+07
              March     2004    1.862076e+06
                        2006    3.527922e+05
              May       2004    1.694353e+06
                        2005   -1.925494e+06
                        2006    2.266991e+04
              November  2004    8.301586e+06
                        2005    4.068098e+05
                        2006    1.108644e+07
              October   2005   -2.605484e+06
                        2006    1.084426e+06
              September 2004    2.041646e+06
                        2005    4.245068e+06
                        2006   -4.177456e+06
Salaried_MNC  April     2004    7.594357e+06
                        2005   -1.432701e+05
                        2006    8.201314e+06
              August    2005   -1.091078e+07
                        2006   -3.434344e+06
              December  2005   -6.249578e+06
                        2006    1.129303e+06
              February  2004    7.479804e+06
                        2005   -2.452103e+07
                        2006   -1.480272e+07
              January   2004    8.534381e+06
                        2005    5.732241e+06
              July      2005    1.434386e+07
                        2006    1.316181e+07
              June      2005   -6.035189e+06
              March     2004   -3.946677e+06
                        2006    1.082340e+07
              May       2004    2.553578e+05
                        2005    3.026666e+07
                        2006    3.573100e+06
              November  2004    6.726835e+06
                        2005    6.736130e+06
                        2006   -4.371106e+06
              October   2005    4.486994e+05
                        2006   -9.080272e+06
              September 2004    4.392743e+06
                        2005   -6.802237e+06
                        2006    4

In [86]:
Profitable_segment.max()

Profit    4.511346e+07
dtype: float64

In [87]:
Profitable_segment.idxmax()

Profit    (Normal Salary, January, 2004)
dtype: object

##  In which category the customers are spending more money?

In [88]:
max_spend_category = merge.groupby(['Type','S Month','S Year'])['Amount'].sum()

Type          S Month    S Year
AIR TICKET    April      2005      1.455950e+07
                         2006      8.662829e+07
              August     2005      2.416482e+07
                         2006      1.352137e+07
              December   2005      1.339723e+07
                                       ...     
TRAIN TICKET  November   2005      6.431938e+07
                         2006      1.199559e+07
              October    2006      1.199578e+05
              September  2004      1.675347e+07
                         2006      3.384603e+07
Name: Amount, Length: 255, dtype: float64

In [89]:
max_spend_category =pd.DataFrame(max_spend_category)
display_full_dataframe(max_spend_category)

Amount
Type         S Month   S Year              
AIR TICKET   April     2005    1.455950e+07
                       2006    8.662829e+07
             August    2005    2.416482e+07
                       2006    1.352137e+07
             December  2005    1.339723e+07
             February  2004    3.821743e+07
                       2005    5.836159e+07
                       2006    3.303725e+07
             January   2004    8.634656e+07
                       2005    1.699645e+07
                       2006    1.093054e+07
             July      2005    3.162036e+06
                       2006    4.200463e+07
             June      2005    4.198977e+07
                       2006    1.915782e+07
             March     2004    1.563774e+06
                       2006    6.405271e+07
             May       2005    9.407585e+07
                       2006    1.002159e+08
             November  2004    3.238930e+07
                       2005    1.851215e+07
             October   2005    1.341386e+07
             September 2005    4.499363e+07
                       2006    4.023098e+06
AUTO         April     2006    1.019220e+07
             December  2005    4.372810e+06
                       2006    1.362008e+07
             February  2005    4.918811e+07
             January   2004    5.571522e+06
             March     2006    9.051983e+07
             May       2005    9.548302e+06
                       2006    1.744640e+07
             November  2006    5.349335e+07
BIKE         April     2005    9.209844e+05
                       2006    1.896219e+07
             February  2004    1.085430e+06
                       2006    6.447005e+07
             January   2004    8.948405e+05
                       2005    5.377784e+07
             March     2006    7.122288e+07
             May       2005    1.001102e+08
             November  2004    5.162799e+05
                       2005    2.212104e+07
                       2006    3.122379e+06
BUS TICKET   April     2004    4.562892e+07
                       2006    5.736882e+06
             August    2005    1.895738e+06
                       2006    1.333180e+07
             February  2004    5.679235e+07
                       2005    6.825012e+07
                       2006    3.916826e+05
             January   2004    1.022591e+08
             July      2006    2.468147e+06
             June      2005    2.772183e+07
                       2006    5.605760e+07
             March     2006    4.950557e+07
             May       2004    2.361539e+06
                       2005    1.282392e+08
             November  2004    1.389835e+07
                       2005    7.759496e+06
                       2006    8.889725e+06
             October   2005    2.956295e+07
             September 2004    1.365688e+07
CAMERA       April     2004    8.875875e+06
                       2005    5.394368e+07
                       2006    5.782767e+07
             August    2005    1.389685e+07
             February  2005    5.792181e+07
                       2006    8.735216e+06
             January   2004    3.661239e+08
                       2005    5.718485e+07
                       2006    1.496501e+07
             July      2005    5.763328e+07
                       2006    4.895010e+07
             June      2005    1.260123e+08
             March     2004    4.375396e+07
                       2006    1.801920e+08
             May       2005    3.700475e+07
                       2006    4.703002e+06
             November  2006    2.300171e+06
             September 2005    8.699794e+06
CAR          April     2006    5.744461e+06
             February  2005    3.948155e+06
                       2006    3.298862e+06
             January   2004    3.339286e+07
             March     2006    3.160803e+07
             May       2006    8.417289e+06
             November  2004    1.144798e+07
             October   2005    2.966197e+06
                       2006    2.

In [90]:
max_spend_category.max()

Amount    3.661239e+08
dtype: float64

In [92]:
max_spend_category.idxmax()

Amount    (CAMERA, January, 2004)
dtype: object

## Impose an interest rate of 2.9% for each customer for any due amount.

In [101]:
positive_profit = merge[merge['Profit']>0]

In [102]:
due_amount = positive_profit.groupby(['S Month','S Year'])['Profit'].sum()

In [104]:
due_amount = pd.DataFrame(due_amount)

In [105]:
display_full_dataframe(due_amount)

Profit
S Month   S Year              
April     2004    5.655164e+07
          2005    7.771466e+07
          2006    2.155404e+08
August    2005    6.627447e+07
          2006    2.166509e+07
December  2005    2.663722e+07
          2006    3.208432e+07
February  2004    6.176561e+07
          2005    3.267080e+08
          2006    9.060878e+07
January   2004    6.194908e+08
          2005    7.363122e+07
          2006    2.638506e+07
July      2005    2.951839e+07
          2006    7.927776e+07
June      2005    1.069655e+08
          2006    4.002931e+07
March     2004    4.117741e+07
          2006    4.192893e+08
May       2004    3.864840e+07
          2005    3.264574e+08
          2006    1.194382e+08
November  2004    3.454360e+07
          2005    9.882112e+07
          2006    5.333834e+07
October   2005    2.411755e+07
          2006    2.553284e+07
September 2004    3.208934e+07
          2005    2.832216e+07
          2006    3.422736e+07

In [106]:
interest_pay = due_amount*0.029

In [107]:
customer_pay = due_amount + interest_pay

In [108]:
customer_pay

Profit
S Month   S Year              
April     2004    5.819164e+07
          2005    7.996838e+07
          2006    2.217911e+08
August    2005    6.819643e+07
          2006    2.229338e+07
December  2005    2.740970e+07
          2006    3.301476e+07
February  2004    6.355681e+07
          2005    3.361826e+08
          2006    9.323644e+07
January   2004    6.374560e+08
          2005    7.576653e+07
          2006    2.715023e+07
July      2005    3.037442e+07
          2006    8.157681e+07
June      2005    1.100675e+08
          2006    4.119016e+07
March     2004    4.237155e+07
          2006    4.314487e+08
May       2004    3.976921e+07
          2005    3.359246e+08
          2006    1.229019e+08
November  2004    3.554536e+07
          2005    1.016869e+08
          2006    5.488515e+07
October   2005    2.481696e+07
          2006    2.627329e+07
September 2004    3.301993e+07
          2005    2.914350e+07
          2006    3.521996e+07

## Monthly profit for the bank.

In [110]:
display_full_dataframe(due_amount)

Profit
S Month   S Year              
April     2004    5.655164e+07
          2005    7.771466e+07
          2006    2.155404e+08
August    2005    6.627447e+07
          2006    2.166509e+07
December  2005    2.663722e+07
          2006    3.208432e+07
February  2004    6.176561e+07
          2005    3.267080e+08
          2006    9.060878e+07
January   2004    6.194908e+08
          2005    7.363122e+07
          2006    2.638506e+07
July      2005    2.951839e+07
          2006    7.927776e+07
June      2005    1.069655e+08
          2006    4.002931e+07
March     2004    4.117741e+07
          2006    4.192893e+08
May       2004    3.864840e+07
          2005    3.264574e+08
          2006    1.194382e+08
November  2004    3.454360e+07
          2005    9.882112e+07
          2006    5.333834e+07
October   2005    2.411755e+07
          2006    2.553284e+07
September 2004    3.208934e+07
          2005    2.832216e+07
          2006    3.422736e+07

# Exporting file

In [115]:
merge.to_excel('task1.xlsx',index=False)